README.rst
sbo-selenium
============

sbo-selenium is a framework primarily intended to help write Selenium tests for
Django applications, although it may be useful in testing other kinds of web
applications as well.

Installation
------------

Add 'sbo-selenium' to the requirements for the application being tested, and add
'sbo_selenium' to the INSTALLED_APPS setting.  Install the new dependency as
usual for the containing project; for SBO packages this is typically::

    pip install -r requirements/tests.txt

Make sure the post-merge runs to install all of the prerequisites::

    ./git-hooks/post-merge

Browser engines
+++++++++++++++

If you are running tests using docker, you don't need to worry about this part.

If you are not using docker, to run tests using a particular browser, the
browser driver needs to be present. To drive Chrome via Selenium, you'll need
to install both Chrome itself and then chromedriver:

1. Download the correct ChromeDriver from
http://chromedriver.storage.googleapis.com/index.html 2. Put the binary,
``chromedriver``, somewhere on your path (ex: ``/usr/local/bin/chromedriver``)

On Mac OS X, if you have Homebrew installed you can instead run
``brew install chromedriver``.

To test Mobile Safari in the iPhone simulator, you'll first need to do the
following:

* Install Xcode
* From the Downloads tab of the Xcode Preferences dialog, install
  "Command Line Tools" and one of the iOS Simulator components (probably the
  latest one)
* Download and run `Appium <http://appium.io/>`_
* Check the "Use Mobile Safari" checkbox
* If you want to test in the iPad form factor, check the "Force Device"
  checkbox and make sure "iPad" is selected next to it
* Click the "Launch" button

To test Opera or Safari, you'll need to download the Selenium standalone server
`jar file <http://selenium-release.storage.googleapis.com/2.40/selenium-server-standalone-2.40.0.jar>`_
and configure the path to it in the ``SELENIUM_JAR_PATH`` setting
described below.

Note that all of these browsers and more can be tested using Sauce OnDemand;
see below for details.

Settings
--------

sbo-selenium uses a few Django settings to configure some aspects of test
runs:

* ``DJANGO_LIVE_TEST_SERVER_ADDRESS`` - The address at which to run the test
  server (this will be used as the environment variable of the same name
  described in the Django testing documentation).  Default value is
  ``'localhost:9090'``.
* ``SELENIUM_DEFAULT_BROWSER`` - The web browser to use for tests when none is
  specified.  Default value is ``'chrome'``.
* ``SELENIUM_DEFAULT_TESTS`` - The Selenium test(s) to be run by default when
  none are specified.  Should be an array of test
  specifications (see `Running Tests`_ below for examples).  Default value is
  an empty list.
* ``SELENIUM_DOCKER_DEBUG`` - True if the debug version of the selenium Docker
  container should be used (in order to allow VNC connections).  Only relevant
  when the ``--docker`` parameter is used.  Default value is ``False``.
* ``SELENIUM_DOCKER_PORT`` - The port of localhost (or the Docker Machine, on
  Mac and Windows) at which to expose the Selenium server running in a Docker
  container.  Only relevant when the ``--docker`` parameter is used.  Default
  value is ``4444``.
* ``SELENIUM_DOCKER_TAG`` - The tag to use in identifying which version of the
  Selenium Docker container to start.  Only relevant when the ``--docker``
  parameter is used.  Default value is ``2.53.0``.
* ``SELENIUM_PAGE_LOAD_TIMEOUT`` - The number of seconds to wait for a response
  to a GET request before considering it to have failed.  Default value is 10
  seconds.  (This is particularly important when using Sauce Connect, as it
  sometimes fails to connect properly, such that every single page load stalls
  indefinitely.)
* ``SELENIUM_JAR_PATH`` - Absolute path of the Selenium standalone server jar
  file.
* ``SELENIUM_POLL_FREQUENCY`` - The number of seconds to wait after a failed
  operation before trying again.  Default value is 0.5 seconds.
* ``SELENIUM_SAUCE_API_KEY`` - The API key for the Sauce Labs account to use
  for running tests.
* ``SELENIUM_SAUCE_CONNECT_PATH`` - Absolute path of the
  `Sauce Connect <https://saucelabs.com/docs/connect>`_ binary, used when
  testing a site that isn't visible to the internet at large via Sauce Labs.
  If not set but other parameters indicate Sauce usage, it's assumed that the
  site being tested is publicly accessible.
* ``SELENIUM_SAUCE_USERNAME`` - The username for the Sauce Labs account to use
  for running tests.
* ``SELENIUM_SAUCE_VERSION`` - The version of Selenium to ask Sauce Labs to
  use in the provided virtual machine.  If omitted, the current default version
  provided by Sauce Labs is used.  Documentation on the available versions and
  current default can be found `here <https://saucelabs.com/docs/additional-config#selenium-version>`_.
* ``SELENIUM_SCREENSHOT_DIR`` - Absolute path of the directory in which to save
  screenshots taken over the course of running tests (these can be useful for
  debugging test failures).  The directory will be created if it doesn't
  already exist.  If the tests are being run via Sauce Labs, screenshots are
  not created in this directory because that service generates screenshots for
  us.
* ``SELENIUM_TEST_COMMAND_OPTIONS`` - Dictionary of additional keyword
  parameters to pass to the ``test`` management command
* ``SELENIUM_TIMEOUT`` - The number of seconds to wait after an operation first
  failed until giving up and declaring it an error.  Default value is 10
  seconds.

Note that some test runners (such as nose) may not always show errors logged
during test setup and teardown.  To consistently get this output, you may want
to configure a logging handler that outputs to file (in addition to any that
may be outputting to console).

Creating Tests
--------------

We're creating Selenium tests as Python classes.  Typically you'll want to
create a common subclass of sbo_selenium.SeleniumTestCase which each of your
application's actual test classes will in turn inherit from.  This main class
will contain methods for common operations like logging in and logging out,
verifying that common page content is present, loading typical test data, and
so forth.  This can be defined in the application's tests.selenium.__init__
module.  Avoid method names with "test" in them, as the test runner is likely
to mistake them for actual standalone tests.

Each related set of tests should then be created as methods in a subclass of
this common test case class which lives in its own file within tests.selenium
(or a subdirectory thereof).  These test methods should start with ``test_`` so
the test runner can find them, and it's generally good practice to have utility
methods which aren't tests themselves start with an underscore.  Typically each
test class will cover functionality on a single page or a set of closely
related pages.

Running Tests
-------------

Running Tests Locally
+++++++++++++++++++++

Tests can be run via the "selenium" management command::

    ./manage.py selenium --settings=myapp.selenium_settings

To run the default set of sbo-selenium tests::

    ./manage.py selenium

Or to specify which test(s) to run rather than using the defaults specified in
the settings file::

    ./manage.py selenium myapp.tests.selenium.test_module:TestClass.test_method myotherapp/tests/selenium --settings=myapp.selenium_settings

Having a separate settings file for the Selenium tests isn't a requirement, but
in practice you'll probably want to use different settings for the tests than
you do for development (for example, to make sure that ``DEBUG=False``).  If you
don't want to type out the settings parameter each time, a simple shell script
should do the trick::

    #!/bin/sh
    ./manage.py selenium $@ --settings=myapp.selenium_settings

All the usual methods that the test runner uses to identify tests should work::

    directory/of/tests
    python.module
    python.module:TestClass
    python.module:TestClass.test_method

(Note that a specifying a package, like myapp.tests.selenium when the actual
tests are defined in modules within that package, does NOT work.)

By default, tests are run in the browser specified by ``SELENIUM_DEFAULT_BROWSER``.
You can use the ``-b`` or ``--browser`` parameter to change this::

    ./manage.py selenium -b firefox
    ./manage.py selenium --browser=safari

Valid browser names are "chrome", "firefox", "htmlunit", "ios", "opera",
"phantomjs", and "safari" ("ipad", "iphone", and "ipod" are treated as
synonyms for "ios", the form factor is chosen in Appium).

If you want to specify a specific Selenium server to use as a command executor
(whether running on localhost, a specific remote server, a Docker container,
etc.), you can use the ``--command-executor`` parameter::

    ./manage.py selenium -b firefox --command-executor=http://127.0.0.1:4444/wd/hub

Note that if you are using a command executor in a Docker container, a remote
host, etc., you should not rely on ``localhost`` or ``127.0.0.1`` in
``DJANGO_LIVE_TEST_SERVER_ADDRESS``.  Instead, either specify the real IP
address (rarely ideal) or try to deduce it via code like the following::

    import socket
    DJANGO_LIVE_TEST_SERVER_ADDRESS = '{}:9090'.format(socket.gethostbyname(socket.gethostname()))

You can also specify the number of times to run the tests (for example, if you
have a test that is failing intermittently for some reason and want to run it
a few times to increase the odds of encountering the error)::

    ./manage.py selenium -n 5

Running Tests in a Docker Container
+++++++++++++++++++++++++++++++++++

As a convenience, you can use the ``--docker`` parameter to automatically start
a standalone Selenium server in a Docker container for chrome or firefox tests.
For this to work, the terminal must already be configured for ``docker``
commands to work.  The container will be stopped automatically at the end of
the test run.  By default it uses the ``2.53.0`` image from
https://hub.docker.com/r/selenium/ and is exposed on port 4444, but this can
be customized via the Django settings described above.::

    ./manage.py selenium --settings=myapp.selenium_settings --docker
    ./manage.py selenium --docker

Running Tests at Sauce Labs
+++++++++++++++++++++++++++

Instead of running tests in a local browser, they can be run on one in a
virtual machine hosted at  `Sauce Labs <https://saucelabs.com/home>`_.  Support
for this in sbo-selenium is designed to play nicely with the Jenkins
`Sauce OnDemand plugin <https://saucelabs.com/jenkins>`_.  For running tests in
Jenkins, just install that plugin and configure it for the job you want to
run.  On a local machine, you'll need to set the ``SELENIUM_SAUCE_*`` Django
settings described above and use a couple of command-line parameters in
addition to the ``-b`` browser name setting mentioned previously:

* ``--platform`` - The name and version of the operating system to
  use.
* ``--browser-version`` - The version number of the browser to use.

The valid values for these can be found on the
`Sauce Labs website <https://saucelabs.com/platforms>`_.

There's also a ``--tunnel-identifier`` parameter which can be used to utilize
a named Sauce Connect tunnel; this is particularly useful if you intend to run
multiple Connect instances against the same account simultaneously (like on
multiple Jenkins slave nodes).  For more about what these tunnels are and when
to use them, see the `Sauce Labs documentation <https://saucelabs.com/docs/connect#part-9>`_
on the topic.

Generating Documentation
------------------------

Documentation for this package is generated using `Sphinx <http://sphinx-doc.org/>`_
and some extensions for it from `sbo-sphinx <https://github.com/safarijv/sbo-sphinx>`_.
To generate the docs locally with any changes you may have made::

    pip install -r requirements/tests.txt
    tox -e docs


docs/CHANGELOG.rst
sbo-selenium Changelog
======================

0.7.3 (2019-02-14)
------------------
* Remove pip as a requirement. (RAMP-146)

0.7.2 (2016-05-05)
------------------
* Clarify and demonstrate how to successfully run tests via a Docker-based
  Selenium server in a virtual machine while using tox.

0.7.1 (2016-04-22)
------------------
* Fixed support for specifying which tests to run on the command line.  This
  seems to have been broken for Django versions which use argparse to parse
  management command options since support was added for them.

0.7.0 (2016-04-21)
------------------
* Added the ``--command-executor`` option for specifying which Selenium server
  to use, if not localhost
* Added the ``--docker`` option to automatically start and stop a Docker
  container for a standalone Selenium server to be used, and added a few
  related Django settings
* Fixed handling of custom parameters to work with recent versions of
  django-nose
* Changed monitoring of the Selenium server auto-launching to watch stderr
  instead of stdout for the startup complete message, as recent versions of
  Selenium seem to have moved the startup output there
* Fixed a bug where the code for timing out attempts to start certain services
  (like a standalone Selenium server or Sauce Connect) was taking far longer
  than intended to time out

0.6.0 (2016-03-29)
------------------
* Removed the ``-p`` alias for ``--platform``, as it now conflicts with
  the ``--plugins`` parameter in the base Django test command
* Fixed support for the underlying test command's standard options under
  Django 1.8 and above
* Confirmed compatibility with Django 1.9 and current master
* Fixed a minor incompatibility in ``audit_accessibility()`` with
  accessibility developer tools 2.10.1

0.5.1 (2016-03-03)
------------------
* Support accessibility developer tools 2.10.1

0.5.0 (2015-06-29)
------------------
* Support Django 1.7 and 1.8
* Support Python 3 when using Django 1.6 and above
* Removed dependency on nose and django-nose, now also works with the default
  Django test runner
* Fixed SELENIUM_DEFAULT_BROWSER behavior (was always defaulting to chrome)
* Added the SELENIUM_TEST_COMMAND_OPTIONS setting

0.4.4 (2015-01-30)
------------------
* Add a hook to allow sub classes to specify a firefox profile (from emperorcezar)
* Fixed compatibility with pip 6.x

0.4.3 (2014-06-01)
------------------
* More robust SeleniumTestCase.click() implementation (retry until success or timeout)

0.4.2 (2014-05-20)
------------------
* Page load timeout support (default is 10 seconds, override via SELENIUM_PAGE_LOAD_TIMEOUT)
* Support for Internet Explorer Sauce OnDemand sessions (from qpfiffer)

0.4.1 (2014-04-18)
------------------
* Added support for Sauce Connect tunnel identifiers
* Added the SELENIUM_SAUCE_VERSION setting to tell Sauce Labs which Selenium
  version to use
* More reliable output of Sauce OnDemand session IDs for integration with
  the Jenkins plugin
* Better redirection of error messages to configured logging (the
  SELENIUM_LOG_FILE setting is no longer needed and has been removed)

0.4.0 (2014-03-29)
------------------
* Initial public release


test_settings.py
import os
import socket

DEBUG = False
JS_DEBUG = False
IP_ADDRESS = socket.gethostbyname(socket.gethostname())
ALLOWED_HOSTS = ['localhost', IP_ADDRESS]
DJANGO_LIVE_TEST_SERVER_ADDRESS = '{}:9090'.format(IP_ADDRESS)
SELENIUM_SAUCE_VERSION = '2.41.0'
SELENIUM_TIMEOUT = 10

# This is a public repository!  Never commit real data for these!
SELENIUM_SAUCE_CONNECT_PATH = ''
SELENIUM_SAUCE_USERNAME = ''
SELENIUM_SAUCE_API_KEY = ''

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# If running in a Windows environment this must be set to the same as your
# system time zone.
TIME_ZONE = 'UTC'

USE_TZ = True

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en-us'

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': ':memory:',
    }
}

ROOT_PATH = os.path.abspath(os.path.dirname(__file__))
LOG_DIR = os.path.join(ROOT_PATH, 'log')

# Absolute path to the directory static files should be collected to.
# Don't put anything in this directory yourself; store your static files
# in apps' "static/" subdirectories and in STATICFILES_DIRS.
# Example: "/home/media/media.lawrence.com/static/"
STATIC_ROOT = os.path.join(ROOT_PATH, 'sbo_selenium', 'static')

# URL prefix for static files.
# Example: "http://media.lawrence.com/static/"
STATIC_URL = '/static/'

# Additional locations of static files
STATICFILES_DIRS = (
    # Put strings here, like "/home/html/static" or "C:/www/django/static".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
)

# List of finder classes that know how to find static files in
# various locations.
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
    # 'django.contrib.staticfiles.finders.DefaultStorageFinder',
)

# List of callables that know how to import templates from various sources.
TEMPLATE_LOADERS = (
    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
    # 'django.template.loaders.eggs.Loader',
)

MIDDLEWARE_CLASSES = (
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    # Uncomment the next line for simple clickjacking protection:
    # 'django.middleware.clickjacking.XFrameOptionsMiddleware',
)

ROOT_URLCONF = 'test_urls'

TEMPLATE_DIRS = (
    # Put strings here, like "/home/html/django_templates" or "C:/www/django/templates".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
)

INSTALLED_APPS = (
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.staticfiles',
    'sbo_selenium',
    'django_nose',
)

try:
    import sbo_sphinx  # flake8: noqa
    INSTALLED_APPS += ('sbo_sphinx',)
except:
    pass

TEST_RUNNER = 'django_nose.NoseTestSuiteRunner'

# This is itself a UUID that is safe to pass to uuid.uuid3
SECRET_KEY = '8ff1e795-3e84-4ef2-b64d-c802b9d4469c'

if not os.path.exists(LOG_DIR):
    os.makedirs(LOG_DIR)
LOGGING = {
    'version': 1,
    'disable_existing_loggers': True,
    'formatters': {
        'simple': {
            'format': '[%(levelname)s] [%(asctime)s] [%(name)s]: %(message)s'
        },
    },
    'handlers': {
        # nose installs handlers that only show output for failed tests, but they
        # can omit errors in setup and teardown; log everything to file also
        'file': {
            'class': 'logging.FileHandler',
            'filename': os.path.join(LOG_DIR, 'tests.log'),
            'formatter': 'simple',
            'level': 'DEBUG',
        },
    },
    'loggers': {
        'django': {
            'handlers': ['file'],
            'level': 'DEBUG',
            'propagate': True,
        },
        'django.request': {
            'handlers': ['file'],
            'level': 'INFO',
            'propagate': True,
        },
        'sbo_selenium': {
            'handlers': ['file'],
            'level': 'DEBUG',
            'propagate': True,
        },
        'selenium': {
            'handlers': ['file'],
            'level': 'DEBUG',
            'propagate': True,
        },
    }
}

NOSE_ARGS = [
    '--nocapture',
    '--failure-detail',
    '--verbosity=3',
]


requirements/clean_up_requirements.py
#!/usr/bin/env python
# encoding: utf-8
# Created by Jeremy Bowman on Mon Jun 29 11:50:12 EDT 2015
# Copyright (c) 2015 Safari Books Online, LLC. All rights reserved.

"""
Utility to uninstall any installed packages listed in uninstall.txt.  Can't use
``pip uninstall -r`` because it aborts after the first package in the list
which isn't currently installed.  Also installs the currently-recommended
versions of setuptools from requirements/base.txt (to avoid needing
to replicate this logic in git-hooks/post-merge, tox.ini, quilter, etc.)
"""

from __future__ import print_function, unicode_literals

import codecs
import os
import re

try:
    from pip.download import PipSession
    from pip.exceptions import UninstallationError
    from pip.index import PackageFinder
    from pip.req import parse_requirements
except ImportError:
    from pip._internal.download import PipSession
    from pip._internal.exceptions import UninstallationError
    from pip._internal.index import PackageFinder
    from pip._internal.req import parse_requirements

if __name__ == '__main__':
    requirements_dir = os.path.abspath(os.path.dirname(__file__))
    requirements_path = os.path.join(requirements_dir, 'uninstall.txt')

    session = PipSession()
    finder = PackageFinder([], [], session=session)
    requirements = parse_requirements(requirements_path, finder, session=session)
    print('Uninstalling former requirements...')
    for req in requirements:
        try:
            req.uninstall(auto_confirm=True)
            print('Uninstalled {}'.format(req.name))
        except UninstallationError as e:
            if 'not installed' in str(e):
                # Good, the whole point was that we don't want it installed
                pass
            else:
                # We probably want to hear about any other uninstallation problems
                raise
    print('Done')

    # Install the currently recommended setuptools
    with codecs.open('requirements/base.txt', 'r', 'utf-8') as f:
        requirements = f.read()
    match = re.search(r'^setuptools==([\d\.]+)$', requirements, re.MULTILINE)
    os.system('pip install setuptools=={}'.format(match.group(1)))


sbo_selenium/management/commands/selenium.py
from __future__ import unicode_literals

from optparse import make_option
import os
from shutil import rmtree
from subprocess import Popen, PIPE

import django
from django.conf import settings as django_settings
from django.core.management import call_command
from django.core.management.base import BaseCommand
from django.test.utils import get_runner

from sbo_selenium.conf import settings
from sbo_selenium.testcase import sauce_sessions
from sbo_selenium.utils import DockerSelenium, OutputMonitor

if 'django_nose' in django_settings.INSTALLED_APPS:
    from django_nose.management.commands.test import Command as TestCommand
else:
    from django.core.management.commands.test import Command as TestCommand

OPTIONS = (
    (('-b', '--browser'), {'dest': 'browser_name', 'default': settings.SELENIUM_DEFAULT_BROWSER, 'help': 'Name of the browser to run the tests in (default is SELENIUM_DEFAULT_BROWSER)'}),
    (('-n',), {'dest': 'count', 'type': int, 'default': 1, 'help': 'Number of times to run each test'}),
    (('--command-executor',), {'dest': 'command_executor', 'help': 'URL of the Selenium server to use if not using Sauce Labs or directly launching a browser'}),
    (('--docker',), {'dest': 'docker', 'action': 'store_true', 'help': 'Start and use a standalone Selenium server in a Docker container (chrome or firefox only)'}),
    (('--platform',), {'dest': 'platform', 'help': 'OS and version thereof for the Sauce OnDemand VM to use'}),
    (('--browser-version',), {'dest': 'browser_version', 'help': 'Browser version for the Sauce OnDemand VM to use'}),
    (('--tunnel-identifier',), {'dest': 'tunnel_id', 'help': 'Sauce Connect tunnel identifier'}),
)


class Command(BaseCommand):
    """
    Django management command for running Selenium tests.
    """
    help = 'Run Selenium tests for this application'
    requires_model_validation = True
    # Command line arguments for Django 1.7 and below
    custom_options = tuple([make_option(*option[0], **option[1]) for option in OPTIONS])

    @property
    def use_argparse(self):
        return not (django.VERSION[0] == 1 and django.VERSION[1] < 8)

    def add_arguments(self, parser):
        """Command line arguments for Django 1.8+"""
        # Add the underlying test command arguments first
        test_command = TestCommand()
        test_command.add_arguments(parser)

        for option in OPTIONS:
            parser.add_argument(*option[0], **option[1])

        if django.VERSION > (1, 9):
            parser.add_argument('args', nargs='*')

    def create_parser(self, prog_name, subcommand):
        """
        Override the base create_parser() method to add this command's custom
        options in Django 1.7 and below.
        """
        if not self.use_argparse:
            self.__class__.option_list = TestCommand.option_list + self.custom_options
        parser = super(Command, self).create_parser(prog_name, subcommand)
        return parser

    @staticmethod
    def clean():
        """Clear out any old screenshots"""
        screenshot_dir = settings.SELENIUM_SCREENSHOT_DIR
        if screenshot_dir and os.path.isdir(screenshot_dir):
            rmtree(screenshot_dir, ignore_errors=True)

    def handle(self, *args, **options):
        """
        Run the specified Selenium test(s) the indicated number of times in
        the specified browser.
        """
        browser_name = options['browser_name']
        count = options['count']
        if len(args) > 0:
            tests = list(args)
        else:
            tests = settings.SELENIUM_DEFAULT_TESTS

        # Kill any orphaned chromedriver processes
        process = Popen(['killall', 'chromedriver'],
                        stderr=open(os.devnull, 'w'))
        process.wait()

        # Clear any old log and screenshots
        self.clean()

        docker = None
        sc_process = None
        selenium_process = None
        if options['docker']:
            if browser_name not in ['chrome', 'firefox']:
                self.stdout.write('Only chrome and firefox can currently be run in a Docker container')
                return
            docker = DockerSelenium(browser=browser_name,
                                    port=settings.SELENIUM_DOCKER_PORT,
                                    tag=settings.SELENIUM_DOCKER_TAG,
                                    debug=settings.SELENIUM_DOCKER_DEBUG)
        elif 'platform' in options and settings.SELENIUM_SAUCE_CONNECT_PATH:
            running, sc_process = self.verify_sauce_connect_is_running(options)
            if not running:
                return
        elif browser_name in ['opera', 'safari']:
            running, selenium_process = self.verify_selenium_server_is_running()
            if not running:
                return
        elif browser_name in ['ipad', 'iphone']:
            if not self.verify_appium_is_running():
                return

        # Make it so django-nose won't have nosetests choke on our parameters
        TestRunner = get_runner(django_settings)
        if hasattr(TestRunner, 'django_opts'):
            for option in OPTIONS:
                TestRunner.django_opts.extend(option[0])

        # Configure and run the tests
        try:
            if docker:
                docker.start()
                options['command_executor'] = docker.command_executor()
            self.update_environment(options)
            self.run_tests(tests, browser_name, count)
        finally:
            # Stop the Selenium Docker container, if running
            if docker and docker.container_id:
                docker.stop()

        # Kill Sauce Connect, if running
        if sc_process:
            sc_process.kill()

        # Kill the Selenium standalone server, if running
        if selenium_process:
            selenium_process.kill()

    def run_tests(self, tests, browser_name, count):
        """Configure and run the tests"""
        test_args = ['test'] + tests
        test_options = settings.SELENIUM_TEST_COMMAND_OPTIONS
        for i in range(count):
            msg = 'Test run %d using %s' % (i + 1, browser_name)
            self.stdout.write(msg)
            call_command(*test_args, **test_options)
            for session in sauce_sessions:
                self.stdout.write(session)
            self.stdout.flush()
            del sauce_sessions[:]

    @staticmethod
    def update_environment(options):
        """
        Populate the environment variables that need to be added for test
        execution to work correctly.  Most (but not all) of these are to match
        what the Jenkins Sauce OnDemand plugin would use for the test
        configuration that was specified on the command line.
        """
        env = os.environ
        # https://docs.djangoproject.com/en/1.6/topics/testing/tools/#liveservertestcase
        env['DJANGO_LIVE_TEST_SERVER_ADDRESS'] = settings.DJANGO_LIVE_TEST_SERVER_ADDRESS
        tunnel_id = options['tunnel_id']
        if tunnel_id:
            env['SAUCE_TUNNEL_ID'] = tunnel_id
        if 'SAUCE_API_KEY' in env:
            # Jenkins plugin has already configured the environment for us
            return
        env['SELENIUM_BROWSER'] = options['browser_name']
        if options['command_executor']:
            env['SELENIUM_COMMAND_EXECUTOR'] = options['command_executor']
        platform = options['platform']
        browser_version = options['browser_version']
        if not platform or not browser_version:
            # None of the following Sauce OnDemand stuff applies
            return
        if settings.SELENIUM_SAUCE_CONNECT_PATH:
            host = 'localhost'
            port = '4445'
        else:
            host = 'ondemand.saucelabs.com'
            port = '80'
        env.update({
            'SAUCE_API_KEY': settings.SELENIUM_SAUCE_API_KEY,
            'SAUCE_USER_NAME': settings.SELENIUM_SAUCE_USERNAME,
            'SELENIUM_BROWSER': options['browser_name'],
            'SELENIUM_HOST': host,
            'SELENIUM_PLATFORM': platform,
            'SELENIUM_PORT': port,
            'SELENIUM_VERSION': browser_version,
        })

    def verify_sauce_connect_is_running(self, options):
        """
        Start Sauce Connect, if it isn't already running.  Returns a tuple of
        two elements:

        * A boolean which is True if Sauce Connect is now running
        * The Popen object representing the process so it can be terminated
          later; if it was already running, this value is "None"
        """
        sc_path = settings.SELENIUM_SAUCE_CONNECT_PATH
        if len(sc_path) < 2:
            self.stdout.write('You need to configure SELENIUM_SAUCE_CONNECT_PATH')
            return False, None
        username = settings.SELENIUM_SAUCE_USERNAME
        if not username:
            self.stdout.write('You need to configure SELENIUM_SAUCE_USERNAME')
            return False, None
        key = settings.SELENIUM_SAUCE_API_KEY
        if not key:
            self.stdout.write('You need to configure SELENIUM_SAUCE_API_KEY')
            return False, None
        # Is it already running?
        process = Popen(['ps -e | grep "%s"' % key],
                        shell=True, stdout=PIPE)
        (grep_output, _grep_error) = process.communicate()
        grep_command = 'grep {}'.format(key)
        lines = grep_output.split('\n')
        for line in lines:
            if 'sc' in line and username in line and grep_command not in line:
                self.stdout.write('Sauce Connect is already running')
                return True, None
        self.stdout.write('Starting Sauce Connect')
        output = OutputMonitor()
        command = [sc_path, '-u', username, '-k', key]
        tunnel_id = options['tunnel_id']
        if tunnel_id:
            command.extend(['-i', tunnel_id])
        sc_process = Popen(command,
                           stdout=output.stream.input,
                           stderr=open(os.devnull, 'w'),
                           universal_newlines=True)
        ready_log_line = 'Connection established.'
        if not output.wait_for(ready_log_line, 60):
            self.stdout.write('Timeout starting Sauce Connect:\n')
            self.stdout.write('\n'.join(output.lines))
            return False, None
        return True, sc_process

    def verify_selenium_server_is_running(self):
        """
        Start the Selenium standalone server, if it isn't already running.
        Returns a tuple of two elements:

        * A boolean which is True if the server is now running
        * The Popen object representing the process so it can be terminated
          later; if the server was already running, this value is "None"
        """
        selenium_jar = settings.SELENIUM_JAR_PATH
        if len(selenium_jar) < 5:
            self.stdout.write('You need to configure SELENIUM_JAR_PATH')
            return False, None
        _jar_dir, jar_name = os.path.split(selenium_jar)
        # Is it already running?
        process = Popen(['ps -e | grep "%s"' % jar_name[:-4]],
                        shell=True, stdout=PIPE)
        (grep_output, _grep_error) = process.communicate()
        lines = grep_output.split('\n')
        for line in lines:
            if jar_name in line:
                self.stdout.write('Selenium standalone server is already running')
                return True, None
        self.stdout.write('Starting the Selenium standalone server')
        output = OutputMonitor()
        selenium_process = Popen(['java', '-jar', selenium_jar],
                                 stdout=open(os.devnull, 'w'),
                                 stderr=output.stream.input)
        ready_log_line = 'Selenium Server is up and running'
        if not output.wait_for(ready_log_line, 10):
            self.stdout.write('Timeout starting the Selenium server:\n')
            self.stdout.write('\n'.join(output.lines))
            return False, None
        return True, selenium_process

    def verify_appium_is_running(self):
        """Verify that Appium is running so it can be used for local iOS tests."""
        process = Popen(['ps -e | grep "Appium"'], shell=True, stdout=PIPE)
        (grep_output, _grep_error) = process.communicate()
        lines = grep_output.split('\n')
        for line in lines:
            if 'Appium.app' in line:
                self.stdout.write('Appium is already running')
                return True
        self.stdout.write('Please launch and configure Appium first')
        return False
